File: aun/aun-sdk-core/go/transport.go
package aun

import (
	"context"
	"crypto/rand"
	"crypto/tls"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"

	"nhooyr.io/websocket"
)

// eventNameMap 协议事件名到 SDK 事件名的映射
var eventNameMap = map[string]string{
	"message.received": "message.received",
	"message.recalled": "message.recalled",
	"message.ack":      "message.ack",
	"group.changed":    "group.changed",
}

// RPCTransport WebSocket JSON-RPC 2.0 传输层
// 与 Python SDK transport.py 对应。
type RPCTransport struct {
	dispatcher    *EventDispatcher
	timeout       time.Duration
	onDisconnect  func(error)
	verifySSL     bool
	ws            *websocket.Conn
	pending       map[string]chan map[string]any
	pendingMu     sync.Mutex
	closed        bool
	closedMu      sync.RWMutex
	challenge     map[string]any
	cancelReader  context.CancelFunc
	readerDone    chan struct{}
}

// NewRPCTransport 创建 RPC 传输层
func NewRPCTransport(dispatcher *EventDispatcher, timeout time.Duration, onDisconnect func(error), verifySSL bool) *RPCTransport {
	if timeout == 0 {
		timeout = 10 * time.Second
	}
	return &RPCTransport{
		dispatcher:   dispatcher,
		timeout:      timeout,
		onDisconnect: onDisconnect,
		verifySSL:    verifySSL,
		pending:      make(map[string]chan map[string]any),
		closed:       true,
	}
}

// SetTimeout 设置 RPC 调用超时时间
func (t *RPCTransport) SetTimeout(timeout time.Duration) {
	t.timeout = timeout
}

// Challenge 返回连接时收到的 challenge 消息
func (t *RPCTransport) Challenge() map[string]any {
	return t.challenge
}

// Connect 连接到 WebSocket 服务端，返回 challenge 消息
func (t *RPCTransport) Connect(ctx context.Context, url string) (map[string]any, error) {
	opts := &websocket.DialOptions{}
	if !t.verifySSL {
		opts.HTTPClient = &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			},
		}
	}
	conn, _, err := websocket.Dial(ctx, url, opts)
	if err != nil {
		return nil, NewConnectionError(fmt.Sprintf("连接 WebSocket 失败: %v", err))
	}

	t.ws = conn
	t.closedMu.Lock()
	t.closed = false
	t.closedMu.Unlock()

	// 接收初始消息（challenge）
	challenge, err := t.recvInitialMessage(ctx)
	if err != nil {
		_ = conn.Close(websocket.StatusNormalClosure, "")
		return nil, err
	}
	t.challenge = challenge

	// 启动读取循环
	readerCtx, cancel := context.WithCancel(context.Background())
	t.cancelReader = cancel
	t.readerDone = make(chan struct{})
	go t.readerLoop(readerCtx)

	return challenge, nil
}

// Close 关闭传输层
func (t *RPCTransport) Close() error {
	t.closedMu.Lock()
	t.closed = true
	t.closedMu.Unlock()

	// 取消读取循环
	if t.cancelReader != nil {
		t.cancelReader()
		// 等待读取循环退出
		if t.readerDone != nil {
			<-t.readerDone
		}
	}

	// 关闭 WebSocket
	if t.ws != nil {
		_ = t.ws.Close(websocket.StatusNormalClosure, "")
		t.ws = nil
	}

	// 通知所有等待中的 RPC 调用
	t.pendingMu.Lock()
	for id, ch := range t.pending {
		close(ch)
		delete(t.pending, id)
	}
	t.pendingMu.Unlock()

	return nil
}

// Call 发起 JSON-RPC 调用
func (t *RPCTransport) Call(ctx context.Context, method string, params map[string]any) (any, error) {
	t.closedMu.RLock()
	if t.closed || t.ws == nil {
		t.closedMu.RUnlock()
		return nil, NewConnectionError("传输层未连接")
	}
	t.closedMu.RUnlock()

	// 生成请求 ID
	rpcID := generateRPCID()
	resultCh := make(chan map[string]any, 1)

	t.pendingMu.Lock()
	t.pending[rpcID] = resultCh
	t.pendingMu.Unlock()

	// 构造 JSON-RPC 请求
	if params == nil {
		params = make(map[string]any)
	}
	request := map[string]any{
		"jsonrpc": "2.0",
		"id":      rpcID,
		"method":  method,
		"params":  params,
	}

	data, err := json.Marshal(request)
	if err != nil {
		t.pendingMu.Lock()
		delete(t.pending, rpcID)
		t.pendingMu.Unlock()
		return nil, NewSerializationError(fmt.Sprintf("序列化 RPC 请求失败: %v", err))
	}

	// 发送请求
	writeCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	if err := t.ws.Write(writeCtx, websocket.MessageText, data); err != nil {
		t.pendingMu.Lock()
		delete(t.pending, rpcID)
		t.pendingMu.Unlock()
		return nil, NewConnectionError(fmt.Sprintf("发送 RPC %s 失败: %v", method, err))
	}

	// 等待响应（带超时）
	timer := time.NewTimer(t.timeout)
	defer timer.Stop()

	select {
	case response, ok := <-resultCh:
		if !ok {
			return nil, NewConnectionError("传输层已关闭")
		}
		// 检查错误
		if errData, ok := response["error"]; ok {
			if errMap, ok := errData.(map[string]any); ok {
				return nil, MapRemoteError(errMap)
			}
		}
		return response["result"], nil

	case <-timer.C:
		t.pendingMu.Lock()
		delete(t.pending, rpcID)
		t.pendingMu.Unlock()
		return nil, NewTimeoutError(fmt.Sprintf("RPC 超时: %s", method))

	case <-ctx.Done():
		t.pendingMu.Lock()
		delete(t.pending, rpcID)
		t.pendingMu.Unlock()
		return nil, NewTimeoutError(fmt.Sprintf("RPC 上下文取消: %s", method))
	}
}

// recvInitialMessage 接收初始消息（通常为 challenge）
func (t *RPCTransport) recvInitialMessage(ctx context.Context) (map[string]any, error) {
	readCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	_, data, err := t.ws.Read(readCtx)
	if err != nil {
		return nil, NewConnectionError(fmt.Sprintf("接收初始消息失败: %v", err))
	}

	message, err := decodeMessage(data)
	if err != nil {
		return nil, err
	}

	if method, ok := message["method"].(string); ok && method == "challenge" {
		return message, nil
	}

	// 非 challenge 消息，路由到事件处理
	t.routeMessage(message)
	return nil, nil
}

// readerLoop 后台读取循环
func (t *RPCTransport) readerLoop(ctx context.Context) {
	defer close(t.readerDone)
	var disconnectErr error
	unexpectedDisconnect := false

	defer func() {
		t.closedMu.Lock()
		wasClosed := t.closed
		t.closed = true
		t.closedMu.Unlock()

		if unexpectedDisconnect && !wasClosed && t.onDisconnect != nil {
			t.onDisconnect(disconnectErr)
		}
	}()

	for {
		t.closedMu.RLock()
		isClosed := t.closed
		t.closedMu.RUnlock()
		if isClosed {
			return
		}

		select {
		case <-ctx.Done():
			return
		default:
		}

		_, data, err := t.ws.Read(ctx)
		if err != nil {
			t.closedMu.RLock()
			wasClosed := t.closed
			t.closedMu.RUnlock()
			if !wasClosed {
				disconnectErr = err
				unexpectedDisconnect = true
				t.dispatcher.Publish("connection.error", map[string]any{"error": err})
			}
			return
		}

		message, err := decodeMessage(data)
		if err != nil {
			log.Printf("解码消息失败: %v", err)
			continue
		}
		t.routeMessage(message)
	}
}

// routeMessage 路由消息到对应的处理器
func (t *RPCTransport) routeMessage(message map[string]any) {
	// 有 id 的是 RPC 响应
	if id, ok := message["id"]; ok {
		rpcID := fmt.Sprintf("%v", id)
		t.pendingMu.Lock()
		ch, exists := t.pending[rpcID]
		if exists {
			delete(t.pending, rpcID)
		}
		t.pendingMu.Unlock()

		if exists {
			ch <- message
		}
		return
	}

	// challenge 消息
	method, _ := message["method"].(string)
	if method == "challenge" {
		t.challenge = message
		params, _ := message["params"].(map[string]any)
		t.dispatcher.Publish("connection.challenge", params)
		return
	}

	// 事件消息
	if len(method) > 6 && method[:6] == "event/" {
		protocolEvent := method[6:]
		sdkEvent := protocolEvent
		if mapped, ok := eventNameMap[protocolEvent]; ok {
			sdkEvent = mapped
		}
		params, _ := message["params"].(map[string]any)
		t.dispatcher.Publish("_raw."+sdkEvent, params)
		return
	}

	// 其他通知
	t.dispatcher.Publish("notification", message)
}

// decodeMessage 解码 WebSocket 消息为 map
func decodeMessage(data []byte) (map[string]any, error) {
	var message map[string]any
	if err := json.Unmarshal(data, &message); err != nil {
		return nil, NewSerializationError("无效的 JSON 载荷")
	}
	return message, nil
}

// generateRPCID 生成随机的 RPC 请求 ID
func generateRPCID() string {
	b := make([]byte, 8)
	_, _ = rand.Read(b)
	return "rpc-" + hex.EncodeToString(b)
}

File: aun/aun-sdk-core/js/src/transport.ts
// ── RPC Transport（浏览器原生 WebSocket）──────────────────

import { EventDispatcher } from './events.js';
import {
  ConnectionError,
  SerializationError,
  TimeoutError,
  mapRemoteError,
} from './errors.js';

/** 协议事件名映射 */
const EVENT_NAME_MAP: Record<string, string> = {
  'message.received': 'message.received',
  'message.recalled': 'message.recalled',
  'message.ack': 'message.ack',
  'group.changed': 'group.changed',
};

/**
 * JSON-RPC 2.0 传输层 — 基于浏览器原生 WebSocket。
 *
 * 职责：
 * - 连接管理（connect / close）
 * - RPC 调用（call — 请求/响应匹配）
 * - 事件路由（服务端推送 → EventDispatcher）
 */
export class RPCTransport {
  private _dispatcher: EventDispatcher;
  private _timeout: number;
  private _onDisconnect: ((error: Error | null) => Promise<void>) | null;
  private _ws: WebSocket | null = null;
  private _closed = true;
  private _challenge: Record<string, unknown> | null = null;
  private _pending: Map<string, {
    resolve: (value: Record<string, unknown>) => void;
    reject: (error: Error) => void;
  }> = new Map();

  constructor(opts: {
    eventDispatcher: EventDispatcher;
    timeout?: number;
    onDisconnect?: ((error: Error | null) => Promise<void>) | null;
  }) {
    this._dispatcher = opts.eventDispatcher;
    this._timeout = opts.timeout ?? 10;
    this._onDisconnect = opts.onDisconnect ?? null;
  }

  /** 设置默认超时（秒） */
  setTimeout(timeout: number): void {
    this._timeout = timeout;
  }

  /** 获取连接时收到的 challenge */
  get challenge(): Record<string, unknown> | null {
    return this._challenge;
  }

  /**
   * 连接到 WebSocket URL。
   * 等待首条消息，若为 challenge 则返回，否则进入消息路由。
   */
  async connect(url: string): Promise<Record<string, unknown> | null> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(url);
      this._ws = ws;
      this._closed = false;
      let initialResolved = false;

      ws.onopen = () => {
        // 连接已建立，等待首条消息
      };

      ws.onerror = (event) => {
        if (!initialResolved) {
          initialResolved = true;
          reject(new ConnectionError(`websocket connection failed to ${url}`));
        }
      };

      ws.onmessage = (event) => {
        if (!initialResolved) {
          // 首条消息：可能是 challenge
          initialResolved = true;
          try {
            const message = this._decodeMessage(event.data);
            if (message.method === 'challenge') {
              this._challenge = message;
              // 切换到正常消息路由
              ws.onmessage = (evt) => this._handleMessage(evt.data);
              ws.onclose = (evt) => this._handleClose(evt);
              ws.onerror = null;
              resolve(message);
            } else {
              // 非 challenge 消息，路由处理后返回 null
              ws.onmessage = (evt) => this._handleMessage(evt.data);
              ws.onclose = (evt) => this._handleClose(evt);
              ws.onerror = null;
              this._routeMessage(message);
              resolve(null);
            }
          } catch (err) {
            reject(err instanceof Error ? err : new ConnectionError(String(err)));
          }
          return;
        }
      };

      ws.onclose = (event) => {
        if (!initialResolved) {
          initialResolved = true;
          reject(new ConnectionError(`websocket closed before initial message: code=${event.code}`));
        }
      };
    });
  }

  /** 关闭连接 */
  async close(): Promise<void> {
    this._closed = true;
    if (this._ws) {
      try {
        this._ws.onclose = null;
        this._ws.onerror = null;
        this._ws.onmessage = null;
        this._ws.close();
      } catch {
        // 忽略关闭异常
      }
      this._ws = null;
    }

    // 拒绝所有待处理的 RPC 调用
    for (const [, pending] of this._pending) {
      pending.reject(new ConnectionError('transport closed'));
    }
    this._pending.clear();
  }

  /**
   * 发起 JSON-RPC 2.0 调用。
   * 返回 result 字段的值；若有 error 字段则抛出映射后的错误。
   */
  async call(
    method: string,
    params?: Record<string, unknown> | null,
    timeout?: number,
  ): Promise<unknown> {
    if (this._closed || !this._ws) {
      throw new ConnectionError('transport not connected');
    }

    // 生成唯一 RPC ID
    const rpcId = `rpc-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 10)}`;
    const effectiveTimeout = (timeout ?? this._timeout) * 1000;

    const promise = new Promise<Record<string, unknown>>((resolve, reject) => {
      this._pending.set(rpcId, { resolve, reject });
    });

    // 发送请求
    try {
      this._ws.send(JSON.stringify({
        jsonrpc: '2.0',
        id: rpcId,
        method,
        params: params ?? {},
      }));
    } catch (exc) {
      this._pending.delete(rpcId);
      throw new ConnectionError(`failed to send rpc ${method}: ${exc}`);
    }

    // 超时控制
    const timeoutPromise = new Promise<never>((_, reject) => {
      globalThis.setTimeout(() => {
        this._pending.delete(rpcId);
        reject(new TimeoutError(`rpc timeout: ${method}`, { retryable: true }));
      }, effectiveTimeout);
    });

    const response = await Promise.race([promise, timeoutPromise]);

    if ('error' in response) {
      throw mapRemoteError(response.error as Record<string, unknown>);
    }
    return response.result;
  }

  // ── 内部消息处理 ──────────────────────────────────

  private _handleMessage(data: unknown): void {
    try {
      const message = this._decodeMessage(data);
      this._routeMessage(message);
    } catch (exc) {
      console.warn('消息解码异常:', exc);
    }
  }

  private _handleClose(event: CloseEvent): void {
    const wasClosed = this._closed;
    this._closed = true;

    // 拒绝所有待处理调用
    for (const [, pending] of this._pending) {
      pending.reject(new ConnectionError(`websocket closed: code=${event.code}`));
    }
    this._pending.clear();

    // 非主动关闭时触发断线回调
    if (!wasClosed) {
      const error = new ConnectionError(`websocket closed: code=${event.code} reason=${event.reason}`);
      this._dispatcher.publish('connection.error', { error });
      if (this._onDisconnect) {
        this._onDisconnect(error).catch(() => {});
      }
    }
  }

  private _routeMessage(message: Record<string, unknown>): void {
    // RPC 响应（有 id 字段）
    if ('id' in message) {
      const rpcId = String(message.id);
      const pending = this._pending.get(rpcId);
      if (pending) {
        this._pending.delete(rpcId);
        pending.resolve(message);
      }
      return;
    }

    const method = String(message.method ?? '');

    // challenge 消息
    if (method === 'challenge') {
      this._challenge = message;
      this._dispatcher.publish('connection.challenge', (message.params ?? {}) as Record<string, unknown>);
      return;
    }

    // 事件推送
    if (method.startsWith('event/')) {
      const protocolEvent = method.slice(6);
      const sdkEvent = EVENT_NAME_MAP[protocolEvent] ?? protocolEvent;
      // 发布为 _raw.{event}，由 AUNClient 处理后再发布用户可见的事件
      this._dispatcher.publish(`_raw.${sdkEvent}`, (message.params ?? {}) as Record<string, unknown>);
      return;
    }

    // 其他通知
    this._dispatcher.publish('notification', message);
  }

  private _decodeMessage(raw: unknown): Record<string, unknown> {
    if (raw !== null && typeof raw === 'object' && !Array.isArray(raw)) {
      return raw as Record<string, unknown>;
    }
    if (typeof raw !== 'string') {
      throw new SerializationError(`unsupported websocket payload type: ${typeof raw}`);
    }
    try {
      return JSON.parse(raw);
    } catch {
      throw new SerializationError('invalid json payload');
    }
  }
}
